### ISSUE TITLE
2.1 Breaking Change for SSL Certificate

### ISSUE THREAD

**Issue by nilobject** [dotnet/corefx#31061]
## Problem: 

A particular integration we have no longer works in 2.1, but it works fine in 2.0. The SSL certificate checks out in CURL, OpenSSL, SSL Labs, and more. When implementing ServerCertificateCustomValidationCallback on a custom HTTP Handler the error is `RemoteCertificateNameMismatch`. However, the name on the URL does match as far as I can tell, and it only is broken in 2.1.

## Environment: 

Docker, running microsoft/dotnet:2.1-runtime on top of Ubuntu and Amazon Linux. Also reproduced without docker on the Ubuntu subsytem on Windows.

## Reproduce Steps

1. `docker run -it microsoft/dotnet:2.1-sdk bash`
2. `dotnet new console -lang 'C#' -n Test`
3. `apt-get update && apt-get install <insert favorite editor>`
4. Edit Test/Program.cs to contain:

```
using System;
using System.Net.Http;
using System.Security.Authentication;

internal class Program
{
  static void Main()
  {
    var client = new HttpClient();
    var resp = client.GetAsync("https://www.ws.xpressreg.net/").Result;
    Console.WriteLine(resp.ToString());
  }
}
```

5. dotnet run

This will generate this output:

```
Unhandled Exception: System.AggregateException: One or more errors occurred. (The SSL connection could not be established, see inner exception.) ---> System.Net.Http.HttpRequestException: The SSL connection could not be established, see inner exception. ---> System.Security.Authentication.AuthenticationException: The remote certificate is invalid according to the validation procedure.
   at System.Net.Security.SslState.StartSendAuthResetSignal(ProtocolToken message, AsyncProtocolRequest asyncRequest, ExceptionDispatchInfo exception)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReceiveBlob(Byte[] buffer, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.CheckCompletionBeforeNextReceive(ProtocolToken message, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartSendBlob(Byte[] incoming, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.ProcessReceivedBlob(Byte[] buffer, Int32 count, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.StartReadFrame(Byte[] buffer, Int32 readBytes, AsyncProtocolRequest asyncRequest)
   at System.Net.Security.SslState.PartialFrameCallback(AsyncProtocolRequest asyncRequest)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Security.SslState.ThrowIfExceptional()
   at System.Net.Security.SslState.InternalEndProcessAuthentication(LazyAsyncResult lazyResult)
   at System.Net.Security.SslState.EndProcessAuthentication(IAsyncResult result)
   at System.Net.Security.SslStream.EndAuthenticateAsClient(IAsyncResult asyncResult)
   at System.Net.Security.SslStream.<>c.<AuthenticateAsClientAsync>b__47_1(IAsyncResult iar)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location where exception was thrown ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.EstablishSslConnectionAsyncCore(Stream stream, SslClientAuthenticationOptions sslOptions, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask)
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at Program.Main() in /Test/Program.cs:line 10
```

If you switch to starting with `docker run -it microsoft/dotnet:2.0-sdk bash` there will be no exception (you'll see a 403 error, which is what I would expect on this bare domain request).

## Investigation Performed

* Tested both www.ws.xpressreg.net and ws.xpressreg.net.
* Tested matching case exactly, e.g., WWW.WS.Xpressreg.Net. Dug into CheckX509HostnameMatch and verified it's using `strncasecmp` so this shouldn't be the issue.
* Tested trying to update ca-certificates, but docker image already has latest version (this was before I knew it was a name mismatch error, not a certificate chain issue)

This code works fine on Mac OS X and Windows with dotnet core 2.1, so it appears to be a Linux specific issue. Since curl works and all SSL utilities I try validate the domain and SSL chain, I have to assume it's not a ca-certificates issue, but is instead localized to the SSL changes made from 2.0 to 2.1.

**Comment by stephentoub** [comment]
I think it's likely this is:
https://github.com/dotnet/corefx/issues/30388
which was fixed recently in the release/2.1 branch by:
https://github.com/dotnet/corefx/pull/30553

cc: @bartonjs, @danmosemsft

**Comment by davidsh** [comment]
The certificate itself uses different casing than the DNS name specified in the code.

Certificate:
![image](https://user-images.githubusercontent.com/7684835/42716025-6e870e3c-86ae-11e8-900d-e3ae78fbc23a.png)

**Comment by nilobject** [comment]
@stephentoub Honestly that looks like exactly it -- when I was guessing on the case sensitivity issue, I looked at the code, but I didn't look at the code as it was tagged for the last release.

@davidsh The RFC states that subject names and alternate names are meant to be treated case insensitively. Also, I did point out that I tried changing the case.

**Comment by bartonjs** [comment]
Yep, seems to just be #30388.  My test tool says it successfully connects using code from master

```
$ ./corerun SslStreamTest.exe -client www.ws.xpressreg.net
Connecting to www.ws.xpressreg.net...
Authenticating connection for 'www.ws.xpressreg.net'...
Certificate subject: CN=WWW.WS.Xpressreg.Net, OU=I.T. Dept., O="Convention Data Services, Inc.", L=Bourne, S=MA, C=US

SSL Policy Errors: None
Chain:
  CN=WWW.WS.Xpressreg.Net, OU=I.T. Dept., O="Convention Data Services, Inc.", L=Bourne, S=MA, C=US
  CN=DigiCert SHA2 High Assurance Server CA, OU=www.digicert.com, O=DigiCert Inc, C=US
  CN=DigiCert High Assurance EV Root CA, OU=www.digicert.com, O=DigiCert Inc, C=US
```

**Comment by davidsh** [comment]
As mentioned, this bug is already fixed in the 2.1 servicing branch.  A servicing update for 2.1 will be shipped with this fix within the next month or two.

Duplicate of #30388
